lib/errors/index.js

```diff
@@ -226,21 +226,96 @@ exports.UnknownConstraintError = UnknownConstraintError;
- * @param {string} type The type of the validation error
- * @param {string} path The field that triggered the validation error
+ * @param {String} message      - An error message
+ * @param {String} type         - The type/origin of the validation error
+ * @param {String} path         - The field that triggered the validation error
+ * @param {String} value        - The value that generated the error
+ * @param {Object} inst         - optional, the DAO instance that caused the validation error
+ * @param {Object} validatorKey - optional, a validation "key", used for identification
+ * @param {String} fnName       - optional, property name of the BUILT-IN validator function that caused the validation error (e.g. "in" or "len"), if applicable
+ * @param {String} fnArgs       - optional, parameters used with the BUILT-IN validator function, if applicable
+ *
  * @param {string} value The value that generated the error
  */
 class ValidationErrorItem {
-  constructor(message, type, path, value) {
+  constructor(message, type, path, value, inst, validatorKey, fnName, fnArgs) {
     this.message = message || '';
-    this.type = type || null;
+    this.type = null;
     this.path = path || null;
     this.value = value !== undefined ? value : null;
-    //This doesn't need captureStackTrace because its not a subclass of Error
+
+    this.origin         = null;
+    this.instance       = inst || null;
+    this.validatorKey   = validatorKey || null;
+    this.validatorName  = fnName || null;
+    this.validatorArgs  = fnArgs || [];
+
+    if (type) {
+      if (ValidationErrorItem.ORIGINS[ type ]) {
+        this.origin = type;
+      } else {
+        const lowercaseType = (type + '').toLowerCase().trim();
+        const realType  = ValidationErrorItem.TypeStringMap[ lowercaseType ];
+
+        if (realType && ValidationErrorItem.ORIGINS[ realType ]) {
+          this.origin = realType;
+          this.type = type;
+        }
+      }
+    }
+  }
+
+  /**
+   * return a lowercase, trimmed string "key" that identifies the validator.
+   *
+   * N.B. the string will be empty if the instance has neither a valid `validatorKey` property nor a valid `validatorName` property
+   *
+   * @param   {Boolean} useTypeAsNS   - optional, defaults to TRUE, controls whether the returned value is "namespace",
+   *                                    this parameter is ignored if the validator's `type` is not one of ValidationErrorItem.ORIGINS
+   * @param   {String}  NSSeparator   - a separator string for concatenating the namespace, must be not be empty,
+   *                                    defaults to "." (fullstop). only used and validated if useTypeAsNS is TRUE.
+   * @throws  {Error}                 - thrown if NSSeparator is found to be invalid.
+   * @return  {String}
+   */
+  getValidatorKey(useTypeAsNS/* = true */, NSSeparator/* = '.'*/) {
+    // parameter defaults: some CI instances give a syntax error on actual parameter default values
```

Comment:
it's simply that Node 4 doesn't support them. You should document the default value in the JSDoc
```js

/**
 * @param {number} [param=10] - this is default param.
 */
function myFunc(param){...}
```